tree/ntuple/v7/src/RNTupleIndex.cxx

```diff
@@ -115,31 +129,45 @@ void ROOT::Experimental::Internal::RNTupleIndex::Build()
 ROOT::Experimental::NTupleSize_t
 ROOT::Experimental::Internal::RNTupleIndex::GetFirstEntryNumber(const std::vector<void *> &valuePtrs) const
 {
-   const auto entryIndices = GetAllEntryNumbers(valuePtrs);
-   if (!entryIndices)
+   const auto entryNumbers = GetAllEntryNumbers(valuePtrs);
+   if (entryNumbers.empty())
       return kInvalidNTupleIndex;
-   return entryIndices->front();
+   return entryNumbers.front();
 }
 
-const std::vector<ROOT::Experimental::NTupleSize_t> *
+const std::vector<ROOT::Experimental::NTupleSize_t>
 ROOT::Experimental::Internal::RNTupleIndex::GetAllEntryNumbers(const std::vector<void *> &valuePtrs) const
 {
    if (valuePtrs.size() != fIndexFields.size())
       throw RException(R__FAIL("Number of value pointers must match number of indexed fields."));
 
    EnsureBuilt();
 
-   std::vector<NTupleIndexValue_t> indexValues;
-   indexValues.reserve(fIndexFields.size());
+   std::vector<std::vector<NTupleIndexValue_t>> entryNumbersPerCluster;
+
+   std::vector<NTupleIndexValue_t> indexFieldValues;
+   indexFieldValues.reserve(fIndexFields.size());
 
    for (unsigned i = 0; i < valuePtrs.size(); ++i) {
-      indexValues.push_back(CastValuePtr(valuePtrs[i], *fIndexFields[i]));
+      indexFieldValues.push_back(CastValuePtr(valuePtrs[i], *fIndexFields[i]));
    }
 
-   auto entryNumber = fIndex.find(RIndexValue(indexValues));
+   RIndexValue indexValue(indexFieldValues);
 
-   if (entryNumber == fIndex.end())
-      return nullptr;
+   for (const auto &indexPartition : fIndexPartitions) {
+      auto clusterEntryNumbers = indexPartition.fIndex.find(indexValue);
+
+      if (clusterEntryNumbers == indexPartition.fIndex.end())
+         continue;
+
+      entryNumbersPerCluster.push_back(clusterEntryNumbers->second);
+   }
```

Comment:
Note that this introduces linear complexity in the number of clusters: while `fIndex.find()` has constant complexity, now there are separate hash maps per cluster. In the end, this means linear complexity in the number of entries (with a small coefficient). Maybe this is fine, but I think this means worse probe performance because of multi-threaded building...